### ISSUE TITLE
ConsoleRunner, nunit2 format xml blank test results

### ISSUE THREAD

**Issue by jakubmichalec** [nunit/nunit-console#290]
I'm upgrading our all automate tests to latest nunit, we have ~1100 selenium tests, during full build some of test results xmls're empty and build fails.

```
$arguments = `"tests.dll`" --test=`"$testCase`" --work=$work --result=nunit-result-$i.xml;format=nunit2 --framework=net-4.5 --noheader --timeout=1500000 --domain=Multiple"

$process = $(Start-Process $nunit $arguments -PassThru)
```

When I use new (nunit3) format all xml's are created as they should.

**Comment by CharliePoole** [comment]
Can you see anything in common in the nunit3 XML for those tests that fail to generate nunit2 results?

**Comment by jakubmichalec** [comment]
nope and even in two almost the same runs (run on weekend so no commits), different tests have empty xmls (I tried to find some pattern for that). If I run much smaller no of test (eq 100), than all work ok.

**Comment by CharliePoole** [comment]
What you're doing in running nunit in a loop is a bit unusual. Could be a re-entrancy problem in the extension. Can you try with a delay between executions?

**Comment by jakubmichalec** [comment]
There's a 100ms delay, I'll increase that

**Comment by edgarssults** [comment]
Did you manage to solve this issue?

We have a similar use case, we have a custom C# application that starts several nunit 3 console runner processes in parallel, each of which executes one unit test (Selenium UI tests). We had issues with Selenium feature context when using nunit 3's parallel execution abilities, which is why we've chosen to run them like this.

And as you describe, sometimes the result XML file is empty. I will experiment with some delays in the code when starting and before disposing of the test runner processes.

Our process arguments in the C# application look like this (process is the console runner executable):
$"\"{testDllPath}\" --where=\"test={testMethodName}\" --result=\"{resultXmlFilePath}\";format=nunit2 --output=\"{outputLogFilePath}\" --workers=1 --agents=1";

**Comment by jakubmichalec** [comment]
@edgarssults nope, I put it on hold atm. I think I need to totaly change our approach, move from 'for' execution to sequence (split tests to smaller groups and than run groups seq in parallel)

**Comment by CharliePoole** [comment]
@jakubmichalec and @edgarssults ...

As I understand it, you want to run various tests defined in the same assembly in parallel groups. I'm not clear if your goal is to run them in separate processes or if you have merely taken that approach because of limitations in NUnit's ability to run them on separate threads in the same process.

Aside from the immediate need you have of making your tests work, it would be helpful to have an issue describing the feature that would allow you to do things in a more straightforward way. Would either of you like to take a shot at that?

**Comment by edgarssults** [comment]
This also looks related to https://github.com/nunit/nunit/issues/2474

I'll try to describe why I'm running each of the tests in a separate process. Not sure whether it should be a feature in NUnit or it's just a problem with another framework that's involved.

We have Selenium tests that are written with the help of Specflow, which generates NUnit unit tests, which then get compiled and executed. This is a normal scenario, but in our case we really wanted the ability for scenarios to execute other scenarios due to the amount of shared test steps we have. So we implemented a step that through reflection finds the feature file and scenario that the user wants and starts its test method. It also copies the current Specflow feature context to the new test instance so that variables are shared.

So far so good, the second scenario will start a new web driver and carry on testing with the previous variables available to it. However, as soon as the second scenario is done, Specflow disposes of something somewhere, I don't remember what. Probably the feature context, which in the Specflow code looked static. It is fine if we are executing one test, but if we are executing several in parallel via NUnit, the disposal messes up the other tests and they fail as soon as the first one finishes. This is what I observed.

The ball is probably in Specflow's court to not make their contexts static. I hope I remembered why exactly I had to execute each tests in its own process, it was a while ago.

**Comment by aolszowka** [comment]
Hi,

We've seen this too, we have several (~20) CI Servers running and we see it most often when the Virtual Machine Host (which hosts most of these CI Servers) is under extreme load (IE something kicked off the build).

We thought it was hardware, but this is the first attempt I've made to actually dig into NUnit source to see what's going on and if there is maybe some type of race condition.

I feel I can add some additional information:
- While the file is "empty" its actually been created by the `CheckWritability(string)` method, I've tested this by building a hacked up nunit-v2-result-writer.dll that has tons of additional logging.
- Therefore it my mind it means that somewhere in between ConsoleRunner.cs 188 and 238 we're dying.

Things I've already tried:
- At first I thought there might be some error in nunit-v2-result-writer with an un-handled exception being tossed so I added a bunch of try/catch with additional logging but came up with nothing.
- I have added additional hacks to log when we enter into an exit out of the `WriteResultFile(XmlNode,string)` method as well as logic around when we enter into `WriteXmlOutput(XmlNode,XmlWriter)` HOWEVER I did not add any logic around `WriteResultFile(XmlNode ,TextWriter)`
- I constantly scour our event logs but nothing ever gets written to them. This runs under CI so anything written to the console is long lost (is there anyway to log that out?)

What I can prove:
- When the error occurs we definitely hit `CheckWritability(string)` (I hacked it up to write out "this is just testing writability" instead of writing an empty file) but at least from what I have seen **we never hit `WriteXmlOutput(XmlNode,XmlWriter)` when we fail.**
- If we do get an exception written to the console its not duplicated to the Event Log. It looks like our process might eat up any error coming out of nunit-console

Here's where I am going next:
- I am going to take a look at our wrapper around nunit-console and make sure we're respecting any non-zero exit code (we might be getting a useful exception but we've lost it???)

Here's how we're running the tests:
`C:\Program Files (x86)\NUnit.org\nunit-console\nunit3-console.exe @C:\BuildOutput\Temp\tmp1042.tmp `

and here is the content of the tmp1042.tmp:
```
C:\BuildOutput\TIMSNET\Release\ComputersUnlimited.DatabaseProgrammability.BusinessIntelligence.SSAS.Tests.dll
--x86
--agents=1
--result=C:\BuildOutput\TIMSNET\Release\ComputersUnlimited.DatabaseProgrammability.BusinessIntelligence.SSAS.Tests.dll.TestResult.xml;format=nunit2
--where=cat==Smoke
```

More background:
These tests come from a time when we did not have parallel abilities, therefore they all need to be run sequentially, however as we've shifted some of our framework towards running in parallel we changed the runner to be NUnit 3 and just change the arguments (so for tests that we can run in parallel our "response file" is much larger with hundreds of test assemblies). But for the legacy stuff we simply call our logic to create a response file with a single test and a single agent and kick off nunit-console.

However this means that we have 100+ nunit-console instances starting up one right after another in very short order. (I'm sorry I know this is terrible, but its what we're dealing with).

We only see this when the Virtual Host is under heavy load (all CI's kick off) so there must be some race condition somewhere that is only getting exposed when the system becomes less responsive.

The tests that seem to fail (its always a different test assembly) are usually extremely small and most times because of the category actually have zero tests to execute, this is probably playing into this (the timing is just way tighter).

I am going to continue digging, but any pointers to where I might go next would be helpful.

**Comment by aolszowka** [comment]
I am sorry for the double post; but I wanted to have this in its own comment:

Sure enough our wrapper was trashing this output:

```
C:\Program Files (x86)\NUnit.org\nunit-console\nunit3-console.exe @C:\BuildOutput\Temp\tmpC883.tmp
NUnit Console Runner 3.8.0
Copyright (c) 2018 Charlie Poole, Rob Prouse
Runtime Environment
OS Version: Microsoft Windows NT 6.3.9600.0
CLR Version: 4.0.30319.42000
Test Files
C:\BuildOutput\TIMSNET\Release\ComputersUnlimited.Tests.TimsProxyInternal.dll
Test Filters
Where: cat==Smoke
System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host
--SocketException
Server stack trace:
at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
at System.Runtime.Remoting.Channels.SocketStream.Read(Byte[] buffer, Int32 offset, Int32 size)
at System.Runtime.Remoting.Channels.SocketHandler.ReadFromSocket(Byte[] buffer, Int32 offset, Int32 count)
at System.Runtime.Remoting.Channels.SocketHandler.Read(Byte[] buffer, Int32 offset, Int32 count)
at System.Runtime.Remoting.Channels.Tcp.TcpFixedLengthReadingStream.Read(Byte[] buffer, Int32 offset, Int32 count)
at System.IO.BinaryReader.ReadBytes(Int32 count)
at System.Runtime.Serialization.Formatters.Binary.SerializationHeaderRecord.Read(__BinaryParser input)
at System.Runtime.Serialization.Formatters.Binary.__BinaryParser.ReadSerializationHeaderRecord()
at System.Runtime.Serialization.Formatters.Binary.__BinaryParser.Run()
at System.Runtime.Serialization.Formatters.Binary.ObjectReader.Deserialize(HeaderHandler handler, __BinaryParser serParser, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize(Stream serializationStream, HeaderHandler handler, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
at System.Runtime.Remoting.Channels.CoreChannel.DeserializeBinaryResponseMessage(Stream inputStream, IMethodCallMessage reqMsg, Boolean bStrictBinding)
at System.Runtime.Remoting.Channels.BinaryClientFormatterSink.SyncProcessMessage(IMessage msg)
Exception rethrown at [0]:
at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
at NUnit.Engine.ITestAgent.Stop()
at NUnit.Engine.Runners.ProcessRunner.Dispose(Boolean disposing)
at NUnit.Engine.Runners.AbstractTestRunner.Dispose()
at NUnit.Engine.Runners.MasterTestRunner.Dispose(Boolean disposing)
at NUnit.Engine.Runners.MasterTestRunner.Dispose()
at NUnit.ConsoleRunner.ConsoleRunner.RunTests(TestPackage package, TestFilter filter)
at NUnit.ConsoleRunner.Program.Main(String[] args)
```

Sure enough we are getting an identical stack trace to that reported in nunit/nunit#2474

To answer the questions that were left unanswered in that original post:

1. framework: NUnit 2.6.4.14350
2. Runner: NUnit3-Console 3.8.0.0 MSI Distribution
3. nunit v2 framework driver extension: The one shipped in the above distribution
4. nunit v2 result writer extension: The one shipped in the above distribution/local hacks to try and target the issue (I think I can remove these hacks going forward) they have reproduced in the past without them.

Other pieces of information that are probably useful (as reported by the NUnit Console output):
- OS Version: Microsoft Windows NT 6.3.9600.0
- CLR Version: 4.0.30319.42000

**Comment by ChrisMaddock** [comment]
Definitely looks similar to me. #372 swallows the exceptions under certain (hopefully, safe!) circumstances - so would hopefully allow the run to complete.

**Comment by aolszowka** [comment]
Thank you, I spoke with the team yesterday, and we aren't hot on taking a nightly binary due to the number of machines involved. Was there an ETA on the 3.9.0 Drop? #390 hasn't been updated in a few weeks, last comment mentions "fall"?

**Comment by ChrisMaddock** [comment]
@rprouse runs the releases of the nunit-console.

I’d definitely recommend trying out the build of master however - more so that, if your issue isn’t resolved, we can try and resolve it in time for the next release! Out of interest - what would be your issue regarding using different versions of the console on different machines?

**Comment by aolszowka** [comment]
@ChrisMaddock Mostly that in order to reproduce this we need to deploy out to 20 different machines and continually kick off the builds to get this to trigger. In the grand scheme of things if our environment was a little more automated for deploy it wouldn't be a big deal.

Is there an MSI Drop somewhere with nightly? Maybe I'll use this as a good time to improve that process.

**Comment by ChrisMaddock** [comment]
There should be an msi available under the appveyor artefacts of the last push to master, if I remember correctly. (Sorry - can’t check on my phone!)

**Comment by jnm2** [comment]
Yes: https://ci.appveyor.com/project/CharliePoole/nunit-console/build/3.9.0-dev-04009/artifacts

**Comment by aolszowka** [comment]
@jnm2 Thank you for the link.

@ChrisMaddock I've got it installed on all of the affected machines (We just recently switched back in January to using NUnit3-Console) so far we have not seen the issue despite attempting to trigger it. We'll run it for a few weeks and get back to you if we see it again.

At very least there were no Regressions on existing tests due to the 3.9.0 Runner Upgrade.

Thank you

**Comment by ChrisMaddock** [comment]
Great, thanks for the update @aolszowka - sounds promising!

**Comment by aolszowka** [comment]
One week of running we have not seen an issue so far (~500 builds have gone through across all 20 servers); this must have been it.

**Comment by ChrisMaddock** [comment]
Great - thanks! Let’s close this issue for now, if anyone else is experiencing the same under he current master build, we can reopen.

**Comment by aolszowka** [comment]
Sorry to be the bearer of bad news, but we encountered this again today (09/21/2018) on one of our CI Servers while under heavy load. This is the first time we've seen it under the 3.9.0 Console Runner. Considering that we've had well over 5,000 Builds since 7/14 across 20 build servers this is unlikely to be reliably reproducible.
